Tests/AmplitudeTests/AmplitudeTests.swift

```diff
@@ -79,43 +79,392 @@ final class AmplitudeTests: XCTestCase {
-        XCTAssertEqual(sessionEndEvent?.sessionId, previousSessionTimestamp)
-        XCTAssertNotNil(sessionEndEvent?.eventId)
+        XCTAssertEqual(sessionEndEvent.eventType, Constants.AMP_SESSION_END_EVENT)
+        XCTAssertEqual(sessionEndEvent.sessionId, previousSessionTimestamp)
+        XCTAssertEqual(sessionEndEvent.timestamp, lastEventTimestamp)
+        XCTAssertEqual(sessionEndEvent.eventId, lastEventId+1)
+
+        XCTAssertNotNil(sessionStartEvent)
+        XCTAssertEqual(sessionStartEvent.eventType, Constants.AMP_SESSION_START_EVENT)
+        XCTAssertEqual(sessionStartEvent.timestamp, currentTimestamp)
+        XCTAssertEqual(sessionStartEvent.sessionId, currentTimestamp)
+        XCTAssertEqual(sessionStartEvent.eventId, lastEventId+2)
+    }
+
+    func testEventSessionsWithTrackingSessionEvents() throws {
+        let lastEventId: Int64 = 123
+        try storageMem.write(key: StorageKey.LAST_EVENT_ID, value: lastEventId)
+
+        let amplitude = Amplitude(configuration: Configuration(
+            apiKey: "testApiKey",
+            storageProvider: storageMem,
+            identifyStorageProvider: interceptStorageMem,
+            minTimeBetweenSessionsMillis: 100,
+            trackingSessionEvents: true
+        ))
+
+        let eventCollector = EventCollectorPlugin()
+        amplitude.add(plugin: eventCollector)
+
+        var currentTimestamp: Int64 = 1000
+        amplitude.track(event: BaseEvent(userId: "user", timestamp: currentTimestamp, eventType: "test event 1"))
+        currentTimestamp = 1050
+        amplitude.track(event: BaseEvent(userId: "user", timestamp: currentTimestamp, eventType: "test event 2"))
+        currentTimestamp = 1200
+        amplitude.track(event: BaseEvent(userId: "user", timestamp: currentTimestamp, eventType: "test event 3"))
+        currentTimestamp = 1350
+        amplitude.track(event: BaseEvent(userId: "user", timestamp: currentTimestamp, eventType: "test event 4"))
+
+        amplitude.onEnterForeground(timestamp: 1500)
+
+        currentTimestamp = 1700
+        amplitude.track(event: BaseEvent(userId: "user", timestamp: currentTimestamp, eventType: "test event 5"))
+
+        amplitude.onExitForeground(timestamp: 1720)
+
+        currentTimestamp = 1750
+        amplitude.track(event: BaseEvent(userId: "user", timestamp: currentTimestamp, eventType: "test event 6"))
+        currentTimestamp = 2000
+        amplitude.track(event: BaseEvent(userId: "user", timestamp: currentTimestamp, eventType: "test event 7"))
+
+        amplitude.onEnterForeground(timestamp: 2050)
+
+        currentTimestamp = 2200
+        amplitude.track(event: BaseEvent(userId: "user", timestamp: currentTimestamp, eventType: "test event 8"))
+
+        amplitude.onExitForeground(timestamp: 2250)
+
+        currentTimestamp = 2400
+        amplitude.track(event: BaseEvent(userId: "user", timestamp: currentTimestamp, eventType: "test event 9"))
+
```

Comment:
Thank you for adding tests @falconandy.

Can you please break this down into smaller more isolated tests to make it easier to follow and so test failures are more specific.
i.e. "testEventsTrackedInBackgroundStartNewSessions", "sessionPersistedWhileInForeground", etc

Reply:
Smaller tests are moved to AmplitudeSessionTests